### PR TITLE
TakeWhile: don't unsubscribe downstream.

### LINKED ISSUE #2647: flatMap doesn't propagate onComplete with unsubscribe of child
The following code hangs forever as it doesn't emit a terminal event:

```java
import java.util.concurrent.TimeUnit;

import rx.Observable;

public class IntervalTakeWhileExample {

    public static void main(String... args) {
        Observable
                .interval(1, TimeUnit.MILLISECONDS)
                .doOnEach(a -> System.out.println("A ==> " + a))
                .takeWhile(aLong -> aLong < 1)
                .doOnEach(b -> System.out.println("B ==> " + b))
                .flatMap(aLong -> {
                    return Observable
                            .timer(30, TimeUnit.MILLISECONDS)
                            .doOnUnsubscribe(() -> System.out.println("unsubscribed timer?"))
                            .doOnEach(c -> System.out.println("C ==> " + c))
                            .map(aLong1 -> aLong1 + aLong);
                })
                .doOnEach(d -> System.out.println("D ==> " + d))
                .toBlocking().forEach(System.out::println);
    }
}
```

This emits:

```
A ==> [rx.Notification@f6adcf11 OnNext 0]
B ==> [rx.Notification@f6adcf11 OnNext 0]
A ==> [rx.Notification@f6adcf12 OnNext 1]
B ==> [rx.Notification@5b2cf4de OnCompleted]
unsubscribed timer?
```

### PR DESCRIPTION
**Pull request by akarnokd**
Fixes #2647 issue with TakeWhile.

### CODE REVIEW

File: src/main/java/rx/Subscriber.java [hunk 1]
```diff
@@ -40,13 +40,24 @@
...
+    /**
+     * Construct a subscriber by using the other subscriber for backpressure
+     * and optionally sharing the underlying subscriptions list.
+     * <p>To retain the chaining of subscribers, the caller should add the
+     * created instance to the op via {@code add()}.
+     * 
+     * @param op the other subscriber
+     * @param shareSubscriptions should the subscription list in op shared with this instance?
+     */
+    protected Subscriber(Subscriber<?> op, boolean shareSubscriptions) {
```
Comment (benjchristensen) [inline src/main/java/rx/Subscriber.java]: Are there other places this constructor should be used where we're doing it manually today?

Comment (akarnokd) [inline src/main/java/rx/Subscriber.java]: Yes, plenty of places.

Comment (benjchristensen) [inline src/main/java/rx/Subscriber.java]: Thought so ... thanks for the confirmation.

Comment (benjchristensen) [inline src/main/java/rx/Subscriber.java]: I just merged this ... but I'm always hesitant when changing something as core as `Subscriber`. Are we ready to support this new method forever. Is it the right signature for all the use cases?

I think it's right, but I've regretted public API decisions before :-)

/cc @zsxwing @abersnaze for more eyes and thought on this.

**Pull request merged**